### PR TITLE
PLT-514: Costing conformance tests

### PR DESCRIPTION
**Pull request by michaelpj**
This adds conformance tests for costing. They are simple additional golden tests.

I also removed some of the conformance test executable that dealt with updating test results. It's much simpler to just use `--accept`, and it's annoying to duplicate the logic.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by michaelpj** [comment]
> This looks fine. Having almost another 500 files containing test output is a bit annoying though. It adds a huge amount of noise to PRs that change them all, and no-one's going to check each one individually. Also, if you do something that makes them change then you're not going to look at the changes: you're just going to type --accept and maybe miss something that you should have paid attention to.

I mean, this is true of all the conformance tests, right? But they _shouldn't_ change unexpectedly ever, really - if we make the conformance tests start failing then we've changed some critical aspect of behaviour!

I guess our tests are now susceptible to the choice of cost model - we probably need to pass it in as a parameter in fact, and if we change the default cost model then the results _will_ change. But then that's expected behaviour.

> Maybe we could do some simple statistical analysis and print out a textual report telling you stuff like how many budgets have increased and decreased, what the average change was and warning you about any outliers.

This seems like a much more relevant suggestion for the _other_ budget tests that we have which test compiler effectiveness and suchlike. Those certainly _do_ change when we fiddle with the compiler and it would be useful to have a better summary of how that looks. But the conformance tests really shouldn't be changing!

File: plutus-conformance/src/PlutusConformance/Common.hs [hunk 1]
```diff
@@ -50,9 +51,16 @@ type UplcProg = UPLC.Program Name DefaultUni DefaultFun ()
 
 -- UPLC evaluation test functions
 
+type UplcEvaluatorFun res = UplcProg -> Maybe res
+
 -- | The evaluator to be tested. It should either return a program if the evaluation is successful,
 -- or `Nothing` if not.
-type UplcEvaluator = UplcProg -> Maybe UplcProg
+type UplcEvaluatorWithoutCosting = UplcEvaluatorFun UplcProg
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-conformance/src/PlutusConformance/Common.hs:58]: Nope, this is just me failing to clean up my dead code from earlier attempts :facepalm:

File: plutus-conformance/executables/test-utils/Spec.hs [hunk 2]
```diff
@@ -51,8 +50,7 @@ dir =
         (metavar "DIR" <> help "The directory the input files are in.")
 
 data Feature
-    = GenTestOutput Runner
-    | Debug
+    = Debug
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-conformance/executables/test-utils/Spec.hs:53]: IMO the whole thing is of questionable utility.

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-conformance/executables/test-utils/Spec.hs:53]: Maybe I should just delete the whole thing

File: plutus-conformance/src/PlutusConformance/Common.hs [hunk 3]
```diff
@@ -176,17 +203,16 @@ updateGoldenFile goldenPath (Right p)  = T.writeFile goldenPath (display p)
...
         -- runCek-like functions (e.g. evaluateCekNoEmit) are partial on term's with free variables,
         -- that is why we manually check first for any free vars
         case UPLC.deBruijnTerm t of
             Left (_ :: UPLC.FreeVariableError) -> Nothing
             Right _                            -> Just ()
-        case evaluateCekNoEmit defaultCekParameters t of
-            Left _     -> Nothing
-            Right prog -> Just prog
+        case runCekNoEmit defaultCekParameters counting t of
+            (Left _, _)                   -> Nothing
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-conformance/src/PlutusConformance/Common.hs:214]: I added a comment.

**Comment by kwxm** [comment]
We could do this with the validation scripts too.  They're a lot more complicated than the conformance tests.

**Comment by michaelpj** [comment]
> We could do this with the validation scripts too. They're a lot more complicated than the conformance tests.

What's "this" here? Add them to the conformance suite? Or add budget tests for them?

**Comment by kwxm** [comment]
> > We could do this with the validation scripts too. They're a lot more complicated than the conformance tests.
> 
> What's "this" here? Add them to the conformance suite? Or add budget tests for them?

I meant that we could add budget tests for them to `plutus-benchmark`.  I suppose we could copy them into `plutus-conformance` too, but I feel slightly resistant to that because they take up a lot of space.  I'm sure that doesn't matter though.  A bigger problem might be that they're `flat` and all the rest are textual, although we could convert them.

**Pull request merged**